pkg/proxy/types.go

```diff
@@ -42,3 +43,33 @@ type ServicePortName struct {
 func (spn ServicePortName) String() string {
 	return fmt.Sprintf("%s:%s", spn.NamespacedName.String(), spn.Port)
 }
+
+// ServiceInfo is an interface which abstracts a service info.
+type ServiceInfo interface {
+	// ClusterIP returns service cluster IP.
+	ClusterIP() string
+	// Port returns service port.
+	Port() int
+	// Protocol returns service protocol.
+	Protocol() api.Protocol
+	// HealthCheckNodePort returns service health check node port.
+	HealthCheckNodePort() int
+}
+
+// EndpointsInfo is an interface which abstracts an endpoints info.
+type EndpointsInfo interface {
```

Comment:
Can we just call this `Endpoint` and rename Endpoint() to String() (and add String() to the Service interface too?

Reply:
Good point!

BTW, there is already a `String()` in `endpointsInfo`, xref: https://github.com/kubernetes/kubernetes/blob/master/pkg/proxy/iptables/proxier.go#L187-L189

I find `endpointsInfo.String()` is never used and can we just change it from 

```
func (e *endpointsInfo) String() string {
	return fmt.Sprintf("%v", *e)
}
```

to 

```
func (e *endpointsInfo) String() string {
	return e.endpoint
}
```

???